pinot-core/src/main/java/org/apache/pinot/core/operator/query/LinearSelectionOrderByOperator.java

```diff
@@ -0,0 +1,452 @@
+    @Nullable
+    private Object[] _lastPartitionRow;
+    private final int _maxSize;
+    private int _alreadySorted;
+
+    public PartiallySortedListBuilder(int maxSize, Comparator<Object[]> partitionComparator,
+        Comparator<Object[]> unsortedComparator) {
+      Preconditions.checkArgument(maxSize > 0, "Max size must be greater than 0");
+      _maxSize = maxSize;
+      _sorted = new ArrayList<>(maxSize);
+      _partitionComparator = partitionComparator;
+      _unsortedComparator = unsortedComparator;
+    }
+
+    @Override
+    public boolean add(Object[] row) {
+      if (_lastPartitionRow == null) {
+        _lastPartitionRow = row;
+        _sorted.add(row);
+        return false;
+      }
+      int cmp = _partitionComparator.compare(row, _lastPartitionRow);
+      if (cmp < 0) {
+        throw new IllegalArgumentException(
+            "Row with docId " + _sorted.size() + " is not sorted compared to the previous one");
+      }
+
+      boolean newPartition = cmp > 0;
+      if (_sorted.size() < _maxSize) { // we don't have _maxSize elements yet
+        if (newPartition) {
+          _lastPartitionRow = row;
+          _alreadySorted = _sorted.size();
+        }
+        // just add the new row to the result list
+        _sorted.add(row);
+        return newPartition;
+      }
+      // at least _maxSize have been added
+      if (newPartition) { // and the new element belongs to a new partition, so we can just ignore it
+        return true;
+      }
+      // new element doesn't belong to a new partition, so we may need to add it
+      if (_lastPartitionQueue == null) { // we have exactly _maxSize elements, and the new belongs to the last partition
+        // we need to prepare the priority queue
+        buildPriorityQueue();
+      }
+      // add the new element if it is lower than the greatest element stored in the partition
+      return false;
+    }
+
+    private void buildPriorityQueue() {
+      int maxPriorityQueueValues = _maxSize - _alreadySorted;
+      _lastPartitionQueue = new PriorityQueue<>(maxPriorityQueueValues, _unsortedComparator.reversed());
+      _lastPartitionQueue.addAll(_sorted.subList(_alreadySorted, _sorted.size()));
+    }
+
+    /**
+     * Sorts the rows and returns the sorted list.
+     *
+     * This object should not be used once this method is called.
+     *
+     * @return the list of all added rows, sorted by {@link #_partitionComparator} and then by
+     * {@link #_unsortedComparator}
+     */
+    @Override
+    public List<Object[]> build() {
+      if (_lastPartitionQueue != null) {
+        assert _lastPartitionQueue.size() == _sorted.size() - _alreadySorted;
+        Iterator<Object[]> lastPartitionIt = _lastPartitionQueue.iterator();
+        for (int i = _alreadySorted; i < _sorted.size(); i++) {
+          _sorted.set(i, lastPartitionIt.next());
+        }
+      }
+      return _sorted;
+    }
+
+    @Override
+    public int sortedSize() {
```

Comment:
Tracking this on the caller side and use it to decide whether to add more rows can be hard to manage. Instead, this logic can be managed internally